pkg/auth/providers/ldap/ldap_actions.go

```diff
@@ -0,0 +1,123 @@
+package ldap
+
+import (
+	"fmt"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/rancher/norman/api/handler"
+	"github.com/rancher/norman/httperror"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/auth/providers/common"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	managementschema "github.com/rancher/types/apis/management.cattle.io/v3/schema"
+	"github.com/rancher/types/apis/management.cattle.io/v3public"
+	"github.com/rancher/types/client/management/v3"
+	"github.com/sirupsen/logrus"
+)
+
+func (p *ldapProvider) formatter(apiContext *types.APIContext, resource *types.RawResource) {
+	common.AddCommonActions(apiContext, resource)
+	resource.AddAction(apiContext, "testAndApply")
+}
+
+func (p *ldapProvider) actionHandler(actionName string, action *types.Action, request *types.APIContext) error {
+	handled, err := common.HandleCommonAction(actionName, action, request, p.providerName, p.authConfigs)
+	if err != nil {
+		return err
+	}
+	if handled {
+		return nil
+	}
+
+	if actionName == "testAndApply" {
+		return p.testAndApply(actionName, action, request)
+	}
+
+	return httperror.NewAPIError(httperror.ActionNotAvailable, "")
+}
+
+func (p *ldapProvider) testAndApply(actionName string, action *types.Action, request *types.APIContext) error {
+	var input map[string]interface{}
+	var err error
+	if p.providerName == "openldap" {
```

Comment:
all these places where you have if statements with openldap vs freeipa, you should instead make the thing that is providers specific a field on the struct and set it when you are creating the struct. So, for example this would just become:
```
input, err = handler.ParseAndValidateActionBody(request, request.Schemas.Schema(&managementschema.Version,
			p.TestAndApplyInputType)
```